WIP: Add utils.symmetric_difference()

frankenjoe opened the pull request:
![image](https://user-images.githubusercontent.com/10383417/181095902-cabb532a-d16d-44ef-becb-1d32eeb0abcf.png)

![image](https://user-images.githubusercontent.com/10383417/181234917-aa333506-0204-4de2-a69c-6c0ffb5f835d.png)

hagenw commented:
Symmetric difference is union without intersection:

![image](https://user-images.githubusercontent.com/173624/181175325-b375006d-aae5-4a5e-b9aa-191c66b14156.png)
([Wikipedia](https://en.wikipedia.org/wiki/Symmetric_difference))

I created code to test for this:

```python
def expected_difference(objs):
    union = audformat.utils.union(objs)
    intersection = audformat.utils.intersect(objs)
    difference = []
    for idx1 in range(len(union)):
        for idx2 in range(len(intersection)):
            if union[idx1] != intersection[idx2]:
                difference.append(union[idx1])
    return pd.MultiIndex.from_tuples(difference, names=union.names)
```

With this your last docstring example
```python
expected_difference(
    [
        audformat.filewise_index(['f1']),
        audformat.filewise_index(['f1', 'f2']),
        audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1]),
    ]
)
```
should return
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

hagenw commented on audformat/core/utils.py:
```diff
@@ -174,7 +175,7 @@ def concat(
             objs = [to_segmented_index(obj) for obj in objs]
     else:
-        _convert_single_level_multi_index(objs)
+        objs = _convert_single_level_multi_index(objs)
```
Interesting, why did the code worked before without any errors?
Because we modified the objects in place?

frankenjoe replied:
The type hint says we have a `Sequence`, which does not support item assignment, i.e. we have to convert the sequence to a list inside the function. And since we now have a new object we must return it.

frankenjoe commented:
> should return
> 
> ```
> MultiIndex([('f1', '0 days',               NaT),
>             ('f2', '0 days',               NaT),
>             ('f2', '0 days', '0 days 00:00:01'),
>             ('f3', '0 days', '0 days 00:00:01')],
>            names=['file', 'start', 'end'])
> ```

We have file 'f1' in the first and second index, so I don't think it should be returned. And the segment `(f1, 0, 1)` we have only in the third index, so I think it should be part of the result.

frankenjoe commented:
Mhh, but when I look at the following results your suggestion indeed makes more sense:

```python
objs = [
        audformat.filewise_index(['f1']),
        audformat.filewise_index(['f1', 'f2']),
        audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1]),
]

index1 = audformat.utils.symmetric_difference(objs)
index2 = audformat.utils.intersect(objs)
index3 = audformat.utils.union([index1, index2])
index4 = audformat.utils.union(objs)

print(index1)
print(index2)
print(index3)
print(index4)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f2', '0 days',               NaT)],
           names=['file', 'start', 'end'])
MultiIndex([('f1', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f2', '0 days',               NaT)],
           names=['file', 'start', 'end'])
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

frankenjoe commented:
So maybe our `intersect()` function is not working is buggy?

hagenw commented:
I think you are right:

```
MultiIndex([('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

should be the result.

For that also (A Δ B) Δ C = A Δ (B Δ C) holds.

My function was wrong and needs to be modified to:

```python
def expected_difference(objs):
    objs = [audformat.utils.to_segmented_index(obj) for obj in objs]
    union = audformat.utils.union(objs)
    intersection = audformat.utils.intersect(objs)
    if len(intersection) == 0:
        return union
    difference = []
    for idx1 in range(len(union)):
        for idx2 in range(len(intersection)):
            if union[idx1] != intersection[idx2]:
                difference.append(union[idx1])
    difference = list(set(difference))
    return pd.MultiIndex.from_tuples(difference, names=union.names)
```

frankenjoe commented:
It's still a bit unfortunate that as soon as we mix filewise and segmented index objects `union(objs)` is not necessarily `union([intersect(objs), symmetric_difference(objs)])`.

frankenjoe commented:
I added tests for the new function. Ready for review now.

hagenw commented:
> It's still a bit unfortunate that as soon as we mix filewise and segmented index objects `union(objs)` is not necessarily the same as `union([intersect(objs), symmetric_difference(objs)])`.

Yes, but I'm not sure if we want to change this.

I think a proper union would be not

```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

but

```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

and the corresponding symmetric difference

```
MultiIndex([('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f1', '0 days 00:00:01', NaT),
           names=['file', 'start', 'end'])
```

frankenjoe commented:
> I think a proper union would be not

Oh yes, that explains it. So the problem is actually `union()`. We could add an optional flag that remove segments if the union also includes the whole file. Then `union([intersect(objs), symmetric_difference(objs)])` should hold.

hagenw commented:
> > I think a proper union would be not
> 
> Oh yes, that explains it. So the problem is actually `union()`. We could add an optional flag to remove segments if the union also includes the whole file. Then `union(objs) == union([intersect(objs), symmetric_difference(objs)])` should hold.

It's not only union, we would in general need to mask filewise times by segments, as I did in my example above.
Then we would have:

**Union**

```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

**Intersection**

```
MultiIndex([('f1', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

**Symmetric Difference**

```
MultiIndex([('f1', '0 days 00:00:01',      NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

But as I said, I don't think we want to have this.

frankenjoe commented:
Right, at the moment we only do this for `intersection()`.

hagenw commented:
There is still a difference between `audformat.utils.symmetric_difference()` and `expected_difference()`.

```python
index1 = audformat.filewise_index(['f1'])
index2 = audformat.filewise_index(['f1', 'f2'])
index3 = audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1])
```

We then get
```python
audformat.utils.symmetric_difference([index1, index2, index3])
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f2', '0 days',               NaT)],
           names=['file', 'start', 'end'])
```python
expected_difference([index1, index2, index3])
```
```
MultiIndex([('f2', '0 days',               NaT),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f1', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
which is all fine, but if we do
```python
index23 = audformat.utils.symmetric_difference([index2, index3])
audformat.utils.symmetric_difference([index1, index23])
```
we get
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT)],
           names=['file', 'start', 'end'])
```
whereas the other approach still leads to the desired result
```python
index23 = expected_difference([index2, index3])
index23 = audformat.utils.set_index_dtypes(index23, {'file': 'string'})
expected_difference([index1, index23])
```
```
MultiIndex([('f2', '0 days',               NaT),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f1', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

frankenjoe commented:
Ok, but I wonder if anway makes more sense if we implement `symmetric_difference()` in the same way as `intersect()`, i.e. we remove segments if the whole file is still in the result or if the whole file is removed:

```python
index1 = audformat.filewise_index(['f1'])
index2 = audformat.filewise_index(['f1', 'f2'])
index3 = audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1])
audformat.symmetric_difference([index1, index2, index3])
```
```
MultiIndex([('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

hagenw commented on tests/test_utils.py:
```diff
@@ -1819,6 +1819,242 @@ def test_set_index_dtypes(index, dtypes, expected):
+        audformat.utils.symmetric_difference(objs),
+        expected,
+    )
+
```
I'm not completely sure yet if (A Δ B) Δ C = A Δ (B Δ C) should also hold with our special filewise/segmented index handling.
But if yes, you could add a test for it here with:
```suggestion
    if len(objs) > 2: 
        pd.testing.assert_index_equal(
            audformat.utils.symmetric_difference(
                [
                    objs[0],
                    audformat.utils.symmetric_difference(objs[1:]),
                ]
            ),
            audformat.utils.symmetric_difference(
                [
                    audformat.utils.symmetric_difference(objs[:-1]),
                    objs[-1],
                ]
            ),
        )

```

frankenjoe replied:
As you already assumed it fails when a combination of filewise and segmented index objects is passed.

hagenw replied:
Yes, but it still looks like a bug in the implementation.

The first `objs` it fails for is:
```python
index1 = audformat.segmented_index(['f1', 'f2'], [0, 0], [1, 1])
index2 = audformat.segmented_index(['f2', 'f3'], [0, 0], [1, 1])
index3 = audformat.filewise_index(['f1', 'f2'])
objs = [
    index1,
    index2,
    index3,
]
```
The result is:
```python
audformat.utils.symmetric_difference([index1, index2, index3])
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
You get the same when doing:
```python
audformat.utils.symmetric_difference(
    [
        audformat.utils.symmetric_difference([index1, index2]),
        index3,
    ]
)
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
but a different result when doing:
```python
audformat.utils.symmetric_difference(
    [
        index1,
        audformat.utils.symmetric_difference([index2, index3]),
    ]
)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

---

But the last result is wrong in my opinion.
You have
```python
index1
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
and
```python
audformat.utils.symmetric_difference([index2, index3])
```
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
if you take the symmetric difference between them it should be
```
MultiIndex([('f1', '0 days',               NaT),
            ('f2', '0 days',               NaT),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
as we said we remove segments if the whole file is present.

frankenjoe replied:
It's because when you do:

```python
audformat.utils.symmetric_difference(
    [
        index1,
        audformat.utils.symmetric_difference([index2, index3]),
    ]
)
```

you are passing only segmented index objects.

frankenjoe replied:
I agree it's not very nice that we have a different behavior now whether you pass a mixture of filewise and segmented vs. only segmented index object. But then we would need to implement a function that removes all segments that lie within another segment first.

frankenjoe replied:
The same issue we have with `intersect)` btw:

```python
objs = [
    audformat.filewise_index(['f1', 'f2']),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```

vs.

```python
objs = [
    audformat.segmented_index(['f1', 'f2']),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', 'NaT')],
           names=['file', 'start', 'end'])
```

Here we would require a utility function that removes segments that include other segments.

hagenw replied:
I think it would be worse the effort as from a user perspective there is not really a difference between
```python
audformat.filewise_index([f1'])
```
and
```python
audformat.segmented_index(['f1'])
```

frankenjoe replied:
Ok, should we then stop here and try to finish #225 without this PR?

hagenw replied:
Basically we need to convert back to filewise if we have a `NaT`. We have `audformat.utils.to_filewise_index()`, but this is doing something completely different ;)

frankenjoe replied:
On the other hand - this might get quite complicated, e.g.:

```python
objs = [
    audformat.segmented_index('f1', 0, 2),
    audformat.segmented_index('f1', 1, 3),
]
audformat.utils.intersect(objs)
```

This currently returns:

```
MultiIndex([], names=['file', 'start', 'end'])
```

but correct would be:

```
MultiIndex([('f1', '0 days 00:00:01', '0 days 00:00:02')],
           names=['file', 'start', 'end'])
```

frankenjoe replied:
But there might be so many special cases and stuff, I wonder if it is really worth the effort...

frankenjoe replied:
My opinion: continue as as is and support only a mixture of filewise + segmented as special case.

frankenjoe replied:
But what indeed seems to be a bug is:

```python
objs = [
    audformat.segmented_index('f1'),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([('f1', '0 days', NaT)],
           names=['file', 'start', 'end'])
```

This should return:

```
MultiIndex([], names=['file', 'start', 'end'])
```

frankenjoe replied:
The other way round it works:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1'),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

hagenw replied:
Sounds like we should add tests for switching the order of `objs` for our utility functions.

frankenjoe replied:
I think it's a problem with `pd.NaT` and that it matches anything.

hagenw replied:
> My opinion: continue as as is and support only a mixture of filewise + segmented as special case.

In order to get the first release out with `MiscTable` this sounds reasonable. But I would open an issue, stating that we should fix this before we make the 1.0.0 release.

frankenjoe replied:
Because this works:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1', 0, 2),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

and:


```python
objs = [
    audformat.segmented_index('f1', 0, 2),
    audformat.segmented_index('f1', 0, 1),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

frankenjoe replied:
> In order to get the first release out with MiscTable this sounds reasonable. But I would open an issue, stating that we should fix this before we make the 1.0.0 release.

Yes, but I would only fix the order bug, whereas the following seems correct to me:

```python
objs = [
    audformat.segmented_index('f1', 0, 1),
    audformat.segmented_index('f1'),
]
audformat.utils.intersect(objs)
```
```
MultiIndex([], names=['file', 'start', 'end'])
```

Cause let's say you have a segmented index and it includes `'f1', 0, 1` and `'f1', 0, pd.NaT`  and then you drop `'f1', 0, pd.NaT`, I don't think you necessarily expect that also `'f1', 0, 1` is dropped. If you a user really wants this, then she can do it by passing a filewise index to explicitly force that all segments belonging to `'f1'` should be removed.

frankenjoe replied:
The correct example would be:

```python
index = audformat.segmented_index(['f1', 'f1'], [0, 0], [1, pd.NaT])
audformat.utils.intersect(
    [
        index,
        audformat.segmented_index('f1'),
    ]
)
```
```
MultiIndex([('f1', '0 days', 'NaT')],
           names=['file', 'start', 'end'])
```

vs.

``` python
index = audformat.segmented_index(['f1', 'f1'], [0, 0], [1, pd.NaT])
audformat.utils.intersect(
    [
        index,
        audformat.filewise_index('f1'),
    ]
)
```
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT)],
           names=['file', 'start', 'end'])
```

hagenw replied:
Which means you would argue the behavior I described above is not a bug, but should be the expected behavior for `symmetric_difference()`, correct?

hagenw replied:
I would also propose to stay with your current implementation here, the question is only we should open a bug and require (A Δ B) Δ C = A Δ (B Δ C). And if not, if we should mention somewhere in the docstrings that this cannot be assumed for a mixture of filewise and segmented indices.

hagenw replied:
> ![image](https://user-images.githubusercontent.com/173624/181281548-1258bd22-9f32-4c3c-8500-5d716c475924.png)

No, we agreed already that we will not handle such overlapping cases. Only the case of `NaT` is considered.
The correct return value is indeed what is returned from the function at the moment.

hagenw commented:
> Ok, but I wonder if anway makes more sense if we implement `symmetric_difference()` in the same way as `intersect()`, i.e. we remove segments if the whole file is in the result or if the whole file is removed:
> 
> ```python
> index1 = audformat.filewise_index(['f1'])
> index2 = audformat.filewise_index(['f1', 'f2'])
> index3 = audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1])
> audformat.symmetric_difference([index1, index2, index3])
> ```
> 
> ```
> MultiIndex([('f2', '0 days',               NaT),
>             ('f3', '0 days', '0 days 00:00:01')],
>            names=['file', 'start', 'end'])
> ```

What is the desired result we need inside `audformat.MiscTable.update()`?

hagenw commented:
We use it at the moment in `drop_index()` with `new_index = self.df.index.difference(index)`.
So, there the desired result if you start with
```python
index = audformat.segmented_index(['f1', 'f2', 'f3'], [0, 0, 0], [1, 1, 1])
```
and want to drop
```python
drop_index = audformat.filewise_index(['f1'])
```
would be indeed
```python
audformat.utils.symmetric_difference([index, drop_index])
```

```
MultiIndex([('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01')],
           names=['file', 'start', 'end'])
```
and not the result we get at the moment, which would be
```
MultiIndex([('f1', '0 days', '0 days 00:00:01'),
            ('f2', '0 days', '0 days 00:00:01'),
            ('f3', '0 days', '0 days 00:00:01'),
            ('f1', '0 days',               NaT)],
           names=['file', 'start', 'end'])
```

So, I think your proposed changes sounds meaningful to me.

frankenjoe commented:
> So, I think your proposed changes sounds meaningful to me.

I changed it accordingly and updated the tests. I added some special case to avoid #230. So we either fix #230 before we merge or we move on and do that later.

frankenjoe commented:
> So we either fix #230 before we merge or we move on and do that later.

I applied the fixed version of `set_index_dtypes()`.

hagenw commented on audformat/core/utils.py:
```diff
@@ -1175,6 +1177,129 @@ def set_index_dtypes(
+    index_intersect = union(pairwise_intersect)
+    index = index_union.difference(index_intersect)
+
+    # index.difference() does not preserve string dtype
+    # for MultiIndex
```
```suggestion
    # index.difference() for MultiIndex
    # does not preserve string and timedelta64[ns] dtypes
```
